### ISSUE TITLE
Support for tomcat form based authentication

### ISSUE THREAD

**Issue by gonzalad** [chexagon/redis-session-manager#14]
I'm using redis-session-manager with Tomcat Form Based Authentication.

When the user submits his credentials, tomcat returns the following error
```
HTTP Status 400 - Invalid direct reference to form login page
```

From what I understand this is because FormAuthenticator (and all tomcat Authentication valves) store session data not in session attributes but in session notes - which is special in-memory hashmap containing non serializable data and as tomcat javadoc says: `This object is <em>not</em> saved and restored across session serializations!`.

The error comes from this line : https://github.com/apache/tomcat80/blob/c2b538c2986c711b9ae949c1f12b70d27783872c/java/org/apache/catalina/authenticator/FormAuthenticator.java#L310
```
        requestURI = savedRequestURL(session);
        if (log.isDebugEnabled()) {
            log.debug("Redirecting to original '" + requestURI + "'");
        }
        if (requestURI == null) {
            if (landingPage == null) {
                response.sendError(HttpServletResponse.SC_BAD_REQUEST,
                        sm.getString("authenticator.formlogin"));
```

landingPage is null because SavedRequest was previously saved in tomcat's session.notes and not in session.attributes.
```
    protected String savedRequestURL(Session session) {

        SavedRequest saved =
            (SavedRequest) session.getNote(Constants.FORM_REQUEST_NOTE);
        if (saved == null) {
            return (null);
        }
```

This is a known limitation of tomcat https://bz.apache.org/bugzilla/show_bug.cgi?id=40531 since 2006 (form base tomcat authentication doesn't work in clustered setup when sticky session is disabled).
Alas tomcat team appears to be unwilling to fix it (see comment in the linked issue : this limitation is a limitation in the servlet spec...).

Perhaps we can either : 
 * try to persuade tomcat team to change their mind about FormAuthenticator behaviour (and all the others authentication valves).
 * implement a specific handling for the Constants.FORM_REQUEST_NOTE and handle java serialization/deserialization of tomcat SavedRequest (non-serializable class) by the plugin itself (hacky but should work).

wdyt ?

Thanks, 
Adrian

**Comment by gonzalad** [comment]
This sample app illustrates the current issue : https://github.com/gonzalad/tomcathasamples/tree/chexagon-issue14/chexagonformauth

**Comment by gonzalad** [comment]
Also a PR is available : #15 which implements solution 2 : 
```
implement a specific handling for the Constants.FORM_REQUEST_NOTE and handle java serialization/deserialization of tomcat SavedRequest (non-serializable class) by the plugin itself (hacky but should work).
```

**Comment by scubasau** [comment]
Closing stale issues
